ddl/cache.lua

```diff
@@ -0,0 +1,110 @@
+local log = require('log')
+local fiber = require('fiber')
+
+local cache = { 
+    spaces = {}, 
+    _initialised = false,
+    _triggers = {},
+    _lock = fiber.channel(1)
+}
+
+local CACHE_LOCK_TIMEOUT  = 3
+local SPACE_NAME_IDX      = 1
+local SHARD_FUNC_NAME_IDX = 2 
+local SHARD_FUNC_BODY_IDX = 3
+
+local function get_metadata(space_name, metadata_name)
+    local metadata_space_name = string.format("_ddl_%s", metadata_name)
+
+    if box.space[metadata_space_name] == nil then
+        return nil
+    end
+
+    return box.space[metadata_space_name]:get{space_name}
+end
+
+local function cache_build()
+    -- Prevent cache_build() from being called concurrently by different fibers.
+    local unlocked = cache._lock:put(true, CACHE_LOCK_TIMEOUT)
+    local trigger_found = false
+
+    if not unlocked then
+        return
+    end
+
+    -- clear cache
+    for key, _ in pairs(cache.spaces) do
+        cache.spaces[key] = nil
+    end
+
+    if box.space._ddl_sharding_func == nil then
+        cache._lock:get()
+        return nil
+    end
+
+    for _, tuple in box.space._ddl_sharding_func:pairs() do
+        cache.spaces[tuple[SPACE_NAME_IDX]] = {}
+        cache.spaces[tuple[SPACE_NAME_IDX]]['sharding_func'] = {}
+
+        cache.spaces[tuple[SPACE_NAME_IDX]]['sharding_func']['raw'] = tuple
+
+        if tuple[SHARD_FUNC_BODY_IDX] ~= nil then
+            local sharding_func, err = loadstring('return ' .. tuple[SHARD_FUNC_BODY_IDX])
+            if sharding_func == nil then
+                cache.spaces[tuple[SPACE_NAME_IDX]]['sharding_func']['processed'] =
+                    string.format("Body is incorrect in sharding_func for space (%s): %s",
+                    tuple[SPACE_NAME_IDX], err)
+            else
+                cache.spaces[tuple[SPACE_NAME_IDX]]['sharding_func']['processed'] =
+                    sharding_func()
+            end
+        elseif tuple[SHARD_FUNC_NAME_IDX] then
+            local chunks = string.split(tuple[SHARD_FUNC_NAME_IDX], '.')
+            cache.spaces[tuple[SPACE_NAME_IDX]]['sharding_func']['processed'] = chunks
+        end
+    end
+
+    cache._schema_version = box.internal.schema_version()
+
+    for _, func in pairs(box.space._ddl_sharding_func:on_replace()) do
+        if func == cache_build then
+            trigger_found = true
+            break
+        end
+    end
+
+    if not trigger_found then
+        box.space._ddl_sharding_func:on_replace(cache_build)
+    end
+
+    cache._initialised = true
+    cache._lock:get()
+end
+
+local function cache_get(space_name, entry)
```

Comment:
It is not clear what `entry` means. Please put a comment for function and describe there arguments and returned values.

Reply:
fixed